### ISSUE TITLE
False Positive of same_functions_in_if_condition with generics

### ISSUE THREAD

**Issue by ModProg** [rust-lang/rust-clippy#10928]
### Summary

when using a function with a generic return, the statements might look more or less identical (in my case they did have different variable names) but they are not.

### Lint Name

clippy::same_functions_in_if_condition

### Reproducer

I tried this code:

```rust
#![warn(clippy::pedantic)]
fn main() {
    let parser = Parser(true);

    if let Some(a) = parser.parse() {
        let _: usize = a;
    } else if let Some(b) = parser.parse() {
        let _: isize = b;
    }
}

struct Parser(bool);

impl Parser {
    fn parse<T: Default>(&self) -> Option<T> {
        self.0.then(T::default)
    }
}
```

I saw this happen:

```
❯ cargo clippy
    Checking sadism v0.1.0 (/mnt/data-ssd/modprog/Development/FOSS/Rust/Play/sadism)
warning: this `if` has the same function call as a previous `if`
 --> src/main.rs:7:15
  |
7 |     } else if let Some(b) = parser.parse() {
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: same as this
 --> src/main.rs:5:8
  |
5 |     if let Some(a) = parser.parse() {
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#same_functions_in_if_condition
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::pedantic)]
  |         ^^^^^^^^^^^^^^^^
  = note: `#[warn(clippy::same_functions_in_if_condition)]` implied by `#[warn(clippy::pedantic)]`

warning: `sadism` (bin "sadism") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
```

I expected to see this happen:

No warning printed as they are not the same, this warning does not happen when the generic is specified on the fn call `.parse::<usize>()`.

### Version

```text
rustc 1.70.0 (90c541806 2023-05-31)
binary: rustc
commit-hash: 90c541806f23a127002de5b4038be731ba1458ca
commit-date: 2023-05-31
host: x86_64-unknown-linux-gnu
release: 1.70.0
LLVM version: 16.0.2
```
